boot: Add the support of MBEDTLS version 3.0.0

Sherryzhang2 opened the pull request:
boot: Add the support of MBEDTLS version 3.0.0 and also keep the backward version support.

Signed-off-by: Sherry Zhang <sherry.zhang2@arm.com>
Change-Id: Idd7ce989fe259e9003732e80beaf3dccdedd3050

davidvincze commented on samples/mcuboot_config/mcuboot_config.template.h:
```diff
@@ -73,6 +73,10 @@
 
 /* Uncomment to use ARM's mbedTLS cryptographic primitives */
 /* #define MCUBOOT_USE_MBED_TLS */
+/* Uncomment to use the MBEDTLS library of version 3.0.0 when
+ * MCUBOOT_USE_MBED_TLS is defined.
+ */
+/* #define MBED_TLS_VERSION_3_0_0 */
```
I would suggest not to define a new macro, please use the macros from the `include/mbedtls/build_info.h` header (used to be in the `include/mbedtls/version.h` header before v3.0). You can use the `MBEDTLS_VERSION_NUMBER` for this purpose. See an example here: `boot/bootutil/src/image_rsa.c:103` .

Sherryzhang2 replied:
Done

d3zd3z commented on boot/bootutil/include/bootutil/crypto/sha256.h:
```diff
@@ -27,6 +28,9 @@
 
 #if defined(MCUBOOT_USE_MBED_TLS)
     #include <mbedtls/sha256.h>
+#if defined (MBED_TLS_VERSION_3_0_0)
```
Given that these are indented, how about indenting the `#if defined` lines added. e.g.
```
#if defined(MCUBOOT_USE_MBED_TLS)
    #include <mbedtls/sha256.h>
    #if defined(MBED_TLS_VERSION_3_0_0)
        #include <mbedtls/compat-2.x.h>
    #endif
    ....
#endif
```

Sherryzhang2 replied:
done

d3zd3z commented on boot/bootutil/include/bootutil/crypto/sha256.h:
```diff
@@ -27,6 +28,9 @@
 
 #if defined(MCUBOOT_USE_MBED_TLS)
     #include <mbedtls/sha256.h>
+#if defined (MBED_TLS_VERSION_3_0_0)
```
Also, there shouldn't be a space between `defined` and the `(`.

Sherryzhang2 replied:
Done

d3zd3z commented on boot/bootutil/include/bootutil/crypto/sha256.h:
```diff
@@ -27,6 +28,10 @@
 #if defined(MCUBOOT_USE_MBED_TLS)
     #include <mbedtls/sha256.h>
+    #include <mbedtls/version.h>
+    #if MBEDTLS_VERSION_NUMBER == 0x03000000
```
We probably want something like `>= 0x03000000` so that it continues to work after there is a new release.

Sherryzhang2 replied:
mbedtls-3.0.0 is the latest tage. Not sure whether it works on future versions.
But should probably work. Will update it.

d3zd3z commented on boot/bootutil/src/image_rsa.c:
```diff
@@ -68,6 +68,13 @@
 #define PSS_MASK_SALT_POS   (PSS_MASK_ONE_POS + 1)
 
+/* Extract a member of the mbedtls_rsa_context structure. */
+#if MBEDTLS_VERSION_NUMBER == 0x03000000
```
Again, >= instead of ==.

Sherryzhang2 replied:
Done

d3zd3z reviewed (changes requested):
Note also, that I've created #1057 to track trying to not depend on private fields.

tamasban commented on boot/bootutil/src/image_rsa.c:
```diff
@@ -88,20 +95,24 @@ bootutil_parse_rsakey(mbedtls_rsa_context *ctx, uint8_t **p, uint8_t *end)
 
-    if ((rc = mbedtls_asn1_get_mpi(p, end, &ctx->N)) != 0 ||
-      (rc = mbedtls_asn1_get_mpi(p, end, &ctx->E)) != 0) {
+    if ((rc = mbedtls_asn1_get_mpi(p, end,
```
In MCUboot there is no 80 char rule. So, might better to keep these are as a single line to improve readability.

Sherryzhang2 replied:
Done

tamasban commented on boot/bootutil/src/image_rsa.c:
```diff
@@ -88,20 +95,24 @@ bootutil_parse_rsakey(mbedtls_rsa_context *ctx, uint8_t **p, uint8_t *end)
     }
 
-    ctx->len = mbedtls_mpi_size(&ctx->N);
+    ctx->MBEDTLS_RSA_CONTEXT_MEMBER(len) = mbedtls_mpi_size(
```
Might better as a single line.

Sherryzhang2 replied:
Done

tamasban commented on boot/bootutil/src/image_rsa.c:
```diff
@@ -112,7 +123,8 @@ bootutil_parse_rsakey(mbedtls_rsa_context *ctx, uint8_t **p, uint8_t *end)
     }
 
-    ctx->len = mbedtls_mpi_size(&ctx->N);
+    ctx->MBEDTLS_RSA_CONTEXT_MEMBER(len) = mbedtls_mpi_size(
```
Might better as a single line.

tamasban reviewed (changes requested):
@d3zd3z: Should not be the internal mbedtls also updated to 3.0.0?  Currently, it is on 2.27.0.

davidvincze reviewed (changes requested):
Please update the `mbedtls_rsa_init()` function's usage in the `src/encrypted.c` file too.
As the migration guide describes:
```
mbedtls_rsa_init(&rsa);
mbedtls_rsa_set_padding(&rsa, MBEDTLS_RSA_PKCS_V21, MBEDTLS_MD_SHA256);
```

The `rsa_alt_helpers.h` header is included from the encrypted.c file, but in 3.0 it has been renamed to `rsa_alt_helpers.h`. Please update the file accordingly. 

Please check the RSA sign and decrypt functions as weel:
https://github.com/ARMmbed/mbedtls/blob/development/docs/3.0-migration-guide.md#the-rng-parameter-is-now-mandatory-for-all-functions-that-accept-one
1 example from encrypted.c  is the mbedtls_rsa_rsaes_oaep_decrypt() function.

davidvincze commented:
> @d3zd3z: Should not be the internal mbedtls also updated to 3.0.0? Currently, it is on 2.27.0.

I agree. Now it passes all the checks, but it only means that in its current state it did not break bacward compatibility.
I new commit should be added on top of `boot: Add the support of MBEDTLS version 3.0.0` so we could test it.

Sherryzhang2 commented:
> > @d3zd3z: Should not be the internal mbedtls also updated to 3.0.0? Currently, it is on 2.27.0.
> 
> I agree. Now it passes all the checks, but it only means that in its current state it did not break bacward compatibility.
> I new commit should be added on top of `boot: Add the support of MBEDTLS version 3.0.0` so we could test it.

I updated the run.c. But I did not figure our how to run the CI with mbedtls-3.0.0.

nvlsianpu commented:
cc @oyvindronningstad, @frkv

davidvincze commented on sim/mcuboot-sys/csupport/run.c:
```diff
@@ -27,6 +27,13 @@
 
```
`mbedtls_rsa_init` needs to be updated in line 151.

Sherryzhang2 replied:
Make sense.

davidvincze commented:
> > > @d3zd3z: Should not be the internal mbedtls also updated to 3.0.0? Currently, it is on 2.27.0.
> > 
> > 
> > I agree. Now it passes all the checks, but it only means that in its current state it did not break bacward compatibility.
> > I new commit should be added on top of `boot: Add the support of MBEDTLS version 3.0.0` so we could test it.
> 
> I updated the run.c. But I did not figure our how to run the CI with mbedtls-3.0.0.

@Sherryzhang2 you have to checkout `mbedtls-3.0.0` version (tag) in the ext/mbedtls submodule and after that you should see in the git status of the mcuboot repo that the submodule has changed:
```
...
(commit or discard the untracked or modified content in submodules)
        modified:   ext/mbedtls (new commits, untracked content)
...
```
After this you add it to your commit. An older example of it: `b04afa9cb`.

Sherryzhang2 commented on sim/mcuboot-sys/csupport/run.c:
```diff
@@ -27,6 +27,13 @@
 
 #define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))
 
+/* Extract a member of the mbedtls_rsa_context structure. */
```
This file only exits in mbedtls-3.0.0. So build time error should happen if include this file directly. It is included implicitly by the mbedtls/***.h.

davidvincze replied:
You are right, leave it as it is. We could include the version.h but it also only includes the build_info.h, doesn't define the version macro even the migration guide says so.

Sherryzhang2 commented on boot/bootutil/include/bootutil/crypto/rsa_alt_helpers.h:
```diff
@@ -0,0 +1,220 @@
+/**
```
Make sense.

Sherryzhang2 commented on boot/bootutil/src/encrypted.c:
```diff
@@ -14,7 +14,11 @@
 #if defined(MCUBOOT_ENCRYPT_RSA)
 #include "mbedtls/rsa.h"
+#if MBEDTLS_VERSION_NUMBER >= 0x03000000
+#include "bootutil/crypto/rsa_alt_helpers.h"
```
Will remove it.

davidvincze commented on boot/bootutil/include/bootutil/crypto/hmac_sha256.h:
```diff
@@ -22,7 +22,9 @@
     #include <mbedtls/cmac.h>
     #include <mbedtls/md.h>
-    #include <mbedtls/md_internal.h>
+    #if MBEDTLS_VERSION_NUMBER < 0x03000000
+        #include <mbedtls/md_internal.h>
+    #endif
```
I think it is okay to completely remove this include from here.
As I see it there has been no reason for it to be included for a while (in hmac_sha256.h).

```suggestion
```

davidvincze commented on boot/bootutil/include/bootutil/crypto/hmac_sha256.h:
```diff
@@ -22,7 +22,9 @@
     #include <mbedtls/cmac.h>
     #include <mbedtls/md.h>
-    #include <mbedtls/md_internal.h>
+    #if MBEDTLS_VERSION_NUMBER < 0x03000000
+        #include <mbedtls/md_internal.h>
+    #endif
```
I think it is okay to completely remove this include from here.
As I see it there has been no reason for it to be included for a while (in hmac_sha256.h).

```suggestion
```

Sherryzhang2 replied:
Done

nvlsianpu commented on boot/bootutil/include/bootutil/crypto/ecdsa_p256.h:
```diff
@@ -34,6 +34,13 @@
+#define MBEDTLS_CONTEXT_MEMBER(X) MBEDTLS_PRIVATE(X)
+#else
+#define MBEDTLS_CONTEXT_MEMBER(X) X
+#endif
```
Shouldn't this be moved to a common header - it is multiplicated in many files.

davidvincze replied:
I agree with you. Currently the relatively best place where we could define it is the `bootutil_priv.h` file, but it would inappropriate there too. I would suggest to simply create a new header, for example: `include/bootutil/crypto/common.h`.

Sherryzhang2 replied:
Done

nvlsianpu commented on boot/bootutil/include/bootutil/crypto/hmac_sha256.h:
```diff
@@ -22,7 +22,6 @@
     #include <stddef.h>
     #include <mbedtls/cmac.h>
     #include <mbedtls/md.h>
-    #include <mbedtls/md_internal.h>
```
is this header inclusion removal backward compatible, or nobody else use this future but mbedtls 3.0.0 users?

nvlsianpu replied:
looks like not a problem

davidvincze replied:
I suggested to remove this include entirely, not just surround it with an `#if MBEDTLS_VERSION_NUMBER` because
there is nothing used in the `hmac_sha256.h` file from the `md_internal.h` header.
I also tested it without upgrading mbedtls in the `ext` folder and all the CI tests passed (no regression according to the CI).

nvlsianpu reviewed (commented):
I'm unable to build mcuboot with ecies-p256 enabled, ASN1 library is incompatible with zephyr's mbedls (signature is RSA using mbedtls). But this was also not possible before this patch.

[edit]: my misunderstood, imgtool said: Error: Signing and encryption must use the same type of key

nvlsianpu reviewed (approved):
I tested a little with zephyr nrf52840dk_target.

Just notice that I need to add sample configuration for x25519 description, as It is impossible to config this properly using menuconfig.

See no regression, no obstacles to merge this PR.

davidvincze reviewed (approved):
LGTM
